notifications/postgres/notificator.go

```diff
@@ -0,0 +1,336 @@
+	queue, err := notifications.NewNotificationQueue(n.queueSize)
+	if err != nil {
+		return nil, invalidRevisionNumber, err
+	}
+	platform := &types.Platform{}
+	err = userContext.Data.Data(platform)
+	if err != nil {
+		return nil, invalidRevisionNumber, fmt.Errorf("could not get platform from user context %v", err)
+	}
+	if platform.ID == "" {
+		return nil, invalidRevisionNumber, errors.New("platform ID not found in user context")
+	}
+	n.isRunningMutex.RLock()
+	defer n.isRunningMutex.RUnlock()
+	if !n.isRunning {
+		return nil, invalidRevisionNumber, errors.New("cannot register consumer - notificator is not running")
+	}
+	if err := n.startListening(); err != nil {
+		return nil, invalidRevisionNumber, fmt.Errorf("listen to %s channel failed %v", postgresChannel, err)
+	}
+
+	n.revisionMutex.RLock()
+	defer n.revisionMutex.RUnlock()
+	n.addConsumer(platform.ID, queue)
+	return queue, n.lastKnownRevision, nil
+}
+
+func (n *notificator) UnregisterConsumer(queue notifications.NotificationQueue) error {
+	n.consumersMutex.Lock()
+	defer n.consumersMutex.Unlock()
+
+	consumerIndex, platformIDToDelete := n.findConsumer(queue.ID())
+	if consumerIndex == -1 {
+		return fmt.Errorf("consumer %s was not found", queue.ID())
+	}
+	platformConsumers := n.consumers[platformIDToDelete]
+	n.consumers[platformIDToDelete] = append(platformConsumers[:consumerIndex], platformConsumers[consumerIndex+1:]...)
+	queue.Close()
+
+	if len(n.consumers[platformIDToDelete]) == 0 {
+		delete(n.consumers, platformIDToDelete)
+	}
+	if len(n.consumers) == 0 {
+		return n.stopListening()
+	}
+	return nil
+}
+
+func (n *notificator) addConsumer(platformID string, queue notifications.NotificationQueue) {
+	n.consumersMutex.Lock()
+	defer n.consumersMutex.Unlock()
+
+	n.consumers[platformID] = append(n.consumers[platformID], queue)
+}
+
+func (n *notificator) findConsumer(id string) (int, string) {
+	var platformIDToDelete string
+	consumerIndex := -1
+	for platformID, platformConsumers := range n.consumers {
+		for index, consumer := range platformConsumers {
+			if consumer.ID() == id {
+				consumerIndex = index
+				break
+			}
+		}
+		if consumerIndex != -1 {
+			platformIDToDelete = platformID
+			break
+		}
+	}
+	return consumerIndex, platformIDToDelete
+}
+
+func (n *notificator) closeAllConsumers() {
+	n.consumersMutex.RLock()
+	defer n.consumersMutex.RUnlock()
+
+	for _, consumers := range n.consumers {
+		for _, consumer := range consumers {
+			consumer.Close()
+		}
+	}
+}
+
+func (n *notificator) setConnection(conn notificationStorage.NotificationConnection) {
+	n.connectionMutex.Lock()
+	defer n.connectionMutex.Unlock()
+	n.connection = conn
+}
+
+func (n *notificator) openConnection() error {
+	connection := n.storage.NewConnection(func(isRunning bool, err error) {
+		n.isRunningMutex.Lock()
+		defer n.isRunningMutex.Unlock()
+		n.isRunning = isRunning
```

Comment:
not sure i get (yet) why is running is necessary - closing the consumers should happen when someone writes to a stop/done channel

Reply:
If the connection to the db is lost then the isRunning is set to false and all current consumers are stopped. You also cannot register new consumers when isRunning is false. When it is true again registration is permitted.